### ISSUE TITLE
Unstable official pangolin rpc node

### ISSUE THREAD

**Issue by HackFisher** [darwinia-network/darwinia-common#1077]
> when our server boots, it fetches all events starting from the block the contract was deployed. i'm trying to look up events in batches of 500 blocks for the last ~10days, which works initially, but eventually i seem to get rate limited/banned before i can grab them all, and the rpc only returns 502 Bad Gateway from there on for each request.
> this would also mean the game isn't available for 10min or so after each app deployment or server restart, whereas other networks are fully initialized in 20-60sec tops, despite running for months already with a whole lot more blocks to look up. going forward, this is becoming an even bigger scaling problem, as the number of blocks to look up is growing steadily.

**Comment by xtools-at** [comment]
hi, i've created a gist to reproduce the issue:
https://gist.github.com/xtools-at/4a1e7d4afd1d2460db8b75d7e3df705a
leave it running for a while and you'll get banned by the RPC, i don't even nearly get close to finishing fetching all relevant events.

what is more, if you try to increase the limit of 1000 blocks, you'll run into "Returned error: query timeout of 10 seconds exceeded". our project uses 3 smart contracts, and fetching all events in batches of 1000s is taking forever. moving forward, this is becoming an even bigger scaling problem, as the number of blocks to look up is growing steadily.

**Comment by xtools-at** [comment]
hi @Jin-Itering, thanks for the extensive feedback, but the answer to "your testnet isn't working for this use case" can't be "deploy on mainnet instead". we have a bunch of beta testers who'd have to jump through hoops to get tokens to pay for their testing efforts, and we deploy contract updates regularly which need to be tested first before they go to mainnet, so there's a need for a stable testing environment.

i've spent quite some time putting the test script above together, and i'm already seeing myself preparing everything and deploying to Crab mainnet, just to run into the same issue after a day's work.
since you guys already boosted the performance of the RPC, it might also be a substrate-related bottleneck in looking up more blocks at a time.

could you use the script above to improve the Pangolin RPCs shortcomings? you could increase the number of blocks to look up on the top (_2000_ blocks make it fail instantly already), and debug the call to the RPC.

**Comment by AsceticBear** [comment]
> since you guys already boosted the performance of the RPC, it might also be a substrate-related bottleneck in looking up more blocks at a time.

Yes, `eth_getLogs` is by far the hardest method to optimize in the standard eth API. The overhead here is the storage access, though some cache has already been added.

**Comment by xtools-at** [comment]
hi @AsceticBear, any updates on this? i'm working on optimising my app in the meantime and pre-caching as many events as possible, but I'd need at least 5k blocks for it to work somewhat stable. The hackathon ends in about a week, would be awesome if my project could still participate.

**Comment by AsceticBear** [comment]
Hi. Sorry for the delay. I had tried to increase the timeout to solve `query timeout of 10 seconds exceeded`. If the `timeout=20s`, then you can fetch 2k blocks event once without the error info above. However, according to the test results of the script you provided, the total fetch time for all the events does not change significantly(I think this approach can not fix your problem). Then, we are trying to add the load balancing mechanism for the pangolin network, it's in progress. 

What's your latest status after your pre-caching events. How long would you wait to fetch all the events to make your application works?

**Comment by AsceticBear** [comment]
I have talked with my DevOps team about the load balancing process. It will be available soon, then we see if that makes the world better. You can track another issue here https://github.com/paritytech/frontier/issues/460 (if interested). I will try to fix it in other ways recently and am glad that you find some workaround solution about this. Yeah. the hackathon is nearly over. You should know that we are aware of this issue with your product, and your product will not be treated differently in the judging process as a result.

**Comment by xtools-at** [comment]
thanks, much appreciated 🙏 subscribed to the new issue

**Comment by HackFisher** [comment]
@xtools-at pangolin's rpc endpoint https://pangolin-rpc.darwinia.network support load balancing now, there are 3-4 nodes running. You can try and see whether  is works.

For the performance root cause of this api, the devs are working on solutions to fix it.

**Comment by AurevoirXavier** [comment]
Any updates?
@AsceticBear

**Comment by AsceticBear** [comment]
I need to handle https://github.com/darwinia-network/darwinia-common/issues/1338 first in next week.
